src/Common/QueryProfiler.cpp

```diff
@@ -40,8 +41,19 @@ namespace
     /// to ignore delivered signals after timer_delete().
     thread_local bool signal_handler_disarmed = true;
 
+    /// Don't permit too many threads be busy inside profiler,
+    /// which could slow down the system in some environments.
+    std::atomic<Int64> concurrent_invocations = 0;
+
     void writeTraceInfo(TraceType trace_type, int /* sig */, siginfo_t * info, void * context)
     {
+        SCOPE_EXIT({ concurrent_invocations.fetch_sub(1, std::memory_order_relaxed); });
+        if (concurrent_invocations.fetch_add(1, std::memory_order_relaxed) > 100)
+        {
+            ProfileEvents::incrementNoTrace(ProfileEvents::QueryProfilerConcurrencyOverruns);
+            return;
```

Comment:
There is a small race. While the thread is doing this increment our counter is higher by 1 than it should be. We can replace it by CAS, but all this will become lock-free instead of wait-free (as it is now), which is not great. The one small thing that we could improve here - is just decrement the counter instantly when we are inside the `if` statement and move SCOPE_EXIT  to `else`.

But CAS is also okay. Maybe even better, because there should not be high contention anyway and it will be 100% correct.

Reply:
I thought about it, and I think the race is ok. It only matters that increments and decrements are done correctly, but it does not matter if the threshold is precise or not. Although it can accidentally overrun, but not much.